REGRESSION (278484@main): Chinese Input Method (pinyin) is broken on ChatGPT/Google translate

whsieh opened the pull request:
#### aa6c1d4264f0176f1a925227bd3e80266ec7e2f6
<pre>
REGRESSION (278484@main): Chinese Input Method (pinyin) is broken on ChatGPT/Google translate
<a href="https://bugs.webkit.org/show_bug.cgi?id=275754">https://bugs.webkit.org/show_bug.cgi?id=275754</a>
<a href="https://rdar.apple.com/130020224">rdar://130020224</a>

Reviewed by Abrar Rahman Protyasha.

After the changes in 278484@main, navigations occurring in subframes that are not empty now cause us
to clear UI-side editor state, through `WebPage::clearEditorStateAfterPageTransition()`. The
intention in 278484@main was to only trigger this codepath when completing a page transition due to
mainframe navigation, but because the `InitializingIframe` flag is set when navigating away from
subframes that have already loaded content, this ended up sometimes clearing the UI-side selection
when loading subframes.

To fix this, this patch begins by reverting the &quot;clear UI-side editor state during page transition&quot;
approach that I previously attempted in 278484@main and 278902@main. Of course, this causes the API
test `ScrollViewScrollabilityTests.ScrollableWithOverflowHiddenWhenZoomed` to start failing again.

After investigating this test failure further, I discovered that this happens because having an up-
to-date `EditorState` with post-layout data means that logic which attempts to adjust the selection
views after scale changes in `-[WKContentView observeValueForKeyPath:ofObject:change:context:]` will
now call into `-_updateChangedSelection:` without returning immediately in that method. As a
consequence, we end up calling into UIKit code under `-selectionChanged`, which (internally)
instantiates the shared `UIKeyboardImpl` instance in the process of posting a notification for
`UITextInputResponderCapabilitiesChangedNotification`.

Instantiating a `UIKeyboardImpl` in the middle of a call to `-[WKScrollView setZoomScale:animated:]`
subsequently causes UIKit to install some native selection views under the scroll view, which (for
reasons that are still unclear) causes the scroll view to revert to a zoom scale of 1.

To keep this test passing after reverting 278484@main and 278902@main, we add a slight optimization
to `-[WKContentView observeValueForKeyPath:ofObject:change:context:]`, such that we avoid trying to
update UIKit&apos;s selection views (and therefore prevent `UIKeyboardImpl` from being instantiated) in
the case where the following conditions are both true:

1. There is no selection (i.e. the current `EditorState` has a selection type that is `None`).
2. No selection was previously shown (i.e. `_lastSelectionDrawingInfo`&apos;s type is also `None`).

There should be no need to update platform selection state in this case, since nothing is visually
changing.

For easier bookkeeping — the full tree of fixes (and followup fixes) is currently:

```
<a href="https://rdar.apple.com/102563363">rdar://102563363</a> (fixed in 276962@main)
    ↳ …caused: <a href="https://rdar.apple.com/126280854">rdar://126280854</a> (fixed in 277910@main)
    ↳ …caused: <a href="https://rdar.apple.com/127547621">rdar://127547621</a> (fixed in 278484@main)
        ↳ …caused: <a href="https://rdar.apple.com/130020224">rdar://130020224</a> (fixed by this PR)
        ↳ …caused: <a href="https://rdar.apple.com/128086114">rdar://128086114</a> (fixed in 278902@main)
            ↳ …caused: <a href="https://rdar.apple.com/128497371">rdar://128497371</a> (fixed by this PR)
```

* LayoutTests/editing/selection/ios/navigating-in-subframe-does-not-dismiss-edit-menu-expected.txt: Added.
* LayoutTests/editing/selection/ios/navigating-in-subframe-does-not-dismiss-edit-menu.html: Added.

Add a new layout test to verify that UI-side editor state isn&apos;t cleared when navigating in subframes
that don&apos;t contain the selection (which is the root cause of the bug on ChatGPT/Google Translate).

* LayoutTests/platform/ios/TestExpectations:

Remove some layout test failure expectations, which are also fixed by this change.

* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/UIProcess/API/gtk/PageClientImpl.h:
* Source/WebKit/UIProcess/API/wpe/PageClientImpl.h:
* Source/WebKit/UIProcess/PageClient.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::clearEditorStateAfterPageTransition): Deleted.

Revert all logic to clear the editor state between page transitions, added in 278484@main.

* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::didClearEditorStateAfterPageTransition): Deleted.
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView observeValueForKeyPath:ofObject:change:context:]):

Apply a mitigation to prevent `ScrollViewScrollabilityTests.ScrollableWithOverflowHiddenWhenZoomed`
from failing again, by preventing unnecessary instantiation of the `UIKeyboardImpl` when changing
the zoom scale.

(-[WKContentView _didClearEditorStateAfterPageTransition]): Deleted.
* Source/WebKit/UIProcess/mac/PageClientImplMac.h:
* Source/WebKit/UIProcess/playstation/PageClientImpl.h:
* Source/WebKit/UIProcess/win/PageClientImpl.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.cpp:
(WebKit::WebLocalFrameLoaderClient::transitionToCommittedForNewPage):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::clearEditorStateAfterPageTransition): Deleted.
* Source/WebKit/WebProcess/WebPage/WebPage.h:

Canonical link: <a href="https://commits.webkit.org/280291@main">https://commits.webkit.org/280291@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/684c3e90d31e290d7a0052ad6106a1630f16d3a0

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/56189 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/35515 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/8661 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/59795 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/6625 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/43137 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/6819 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/45427 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/4600 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/58218 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/33405 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/48472 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/26365 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/55715 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/30185 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/5802 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/5629 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/52185 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/6074 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/61478 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/97 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/6199 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/52751 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/97 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/48539 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/52577 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12449 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/86 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/31342 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/32428 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/33511 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/32175 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on LayoutTests/editing/selection/ios/navigating-in-subframe-does-not-dismiss-edit-menu.html:
```diff
@@ -0,0 +1,69 @@
+    await UIHelper.waitForMenuToShow();
+
+    subframe.srcdoc = `<p>subframe after reloading</p>`;
+    await UIHelper.delayFor(800);
```
The 800ms delay doesn't have a specific meaning right? Is it possible to make it shorter and/or wait for something specific like a presentation update? 😬

whsieh replied:
So this delay allows the edit menu to dismiss, in the case where the bug reproduces (and the editor state is erroneously cleared in the UI process as a result of subframe navigation). It's tricky to avoid _some_ hard-coded delay here, since the test needs to check that the edit menu _doesn't_ dismiss.

aprotyas reviewed (approved):
Looks good! One editorial comment:

> There should be need to update platform selection state in this case, since nothing is visually
changing.

I think you mean "_There should be no need to update_" here instead?

whsieh committed:
REGRESSION (278484@main): Chinese Input Method (pinyin) is broken on ChatGPT/Google translate
https://bugs.webkit.org/show_bug.cgi?id=275754
rdar://130020224

Reviewed by Abrar Rahman Protyasha.

After the changes in 278484@main, navigations occurring in subframes that are not empty now cause us
to clear UI-side editor state, through `WebPage::clearEditorStateAfterPageTransition()`. The
intention in 278484@main was to only trigger this codepath when completing a page transition due to
mainframe navigation, but because the `InitializingIframe` flag is set when navigating away from
subframes that have already loaded content, this ended up sometimes clearing the UI-side selection
when loading subframes.

To fix this, this patch begins by reverting the "clear UI-side editor state during page transition"
approach that I previously attempted in 278484@main and 278902@main. Of course, this causes the API
test `ScrollViewScrollabilityTests.ScrollableWithOverflowHiddenWhenZoomed` to start failing again.

After investigating this test failure further, I discovered that this happens because having an up-
to-date `EditorState` with post-layout data means that logic which attempts to adjust the selection
views after scale changes in `-[WKContentView observeValueForKeyPath:ofObject:change:context:]` will
now call into `-_updateChangedSelection:` without returning immediately in that method. As a
consequence, we end up calling into UIKit code under `-selectionChanged`, which (internally)
instantiates the shared `UIKeyboardImpl` instance in the process of posting a notification for
`UITextInputResponderCapabilitiesChangedNotification`.

Instantiating a `UIKeyboardImpl` in the middle of a call to `-[WKScrollView setZoomScale:animated:]`
subsequently causes UIKit to install some native selection views under the scroll view, which (for
reasons that are still unclear) causes the scroll view to revert to a zoom scale of 1.

To keep this test passing after reverting 278484@main and 278902@main, we add a slight optimization
to `-[WKContentView observeValueForKeyPath:ofObject:change:context:]`, such that we avoid trying to
update UIKit's selection views (and therefore prevent `UIKeyboardImpl` from being instantiated) in
the case where the following conditions are both true:

1. There is no selection (i.e. the current `EditorState` has a selection type that is `None`).
2. No selection was previously shown (i.e. `_lastSelectionDrawingInfo`'s type is also `None`).

There should be no need to update platform selection state in this case, since nothing is visually
changing.

For easier bookkeeping — the full tree of fixes (and followup fixes) is currently:

```
rdar://102563363 (fixed in 276962@main)
    ↳ …caused: rdar://126280854 (fixed in 277910@main)
    ↳ …caused: rdar://127547621 (fixed in 278484@main)
        ↳ …caused: rdar://130020224 (fixed by this PR)
        ↳ …caused: rdar://128086114 (fixed in 278902@main)
            ↳ …caused: rdar://128497371 (fixed by this PR)
```

* LayoutTests/editing/selection/ios/navigating-in-subframe-does-not-dismiss-edit-menu-expected.txt: Added.
* LayoutTests/editing/selection/ios/navigating-in-subframe-does-not-dismiss-edit-menu.html: Added.

Add a new layout test to verify that UI-side editor state isn't cleared when navigating in subframes
that don't contain the selection (which is the root cause of the bug on ChatGPT/Google Translate).

* LayoutTests/platform/ios/TestExpectations:

Remove some layout test failure expectations, which are also fixed by this change.

* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/UIProcess/API/gtk/PageClientImpl.h:
* Source/WebKit/UIProcess/API/wpe/PageClientImpl.h:
* Source/WebKit/UIProcess/PageClient.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::clearEditorStateAfterPageTransition): Deleted.

Revert all logic to clear the editor state between page transitions, added in 278484@main.

* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::didClearEditorStateAfterPageTransition): Deleted.
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView observeValueForKeyPath:ofObject:change:context:]):

Apply a mitigation to prevent `ScrollViewScrollabilityTests.ScrollableWithOverflowHiddenWhenZoomed`
from failing again, by preventing unnecessary instantiation of the `UIKeyboardImpl` when changing
the zoom scale.

(-[WKContentView _didClearEditorStateAfterPageTransition]): Deleted.
* Source/WebKit/UIProcess/mac/PageClientImplMac.h:
* Source/WebKit/UIProcess/playstation/PageClientImpl.h:
* Source/WebKit/UIProcess/win/PageClientImpl.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.cpp:
(WebKit::WebLocalFrameLoaderClient::transitionToCommittedForNewPage):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::clearEditorStateAfterPageTransition): Deleted.
* Source/WebKit/WebProcess/WebPage/WebPage.h:

Canonical link: https://commits.webkit.org/280291@main
